File: javascript/wcs_math.js
//============================================================================
// wcs_math.js - WCS Math Library
//
// Provides TAN (gnomonic) projection, CD matrix fitting, and centroid computation.
// Pure JavaScript compatible with both PJSR and Node.js.
//
// Copyright (c) 2024-2025 Split Image Solver Project
//============================================================================

// Math is available by default in both PJSR and Node.js environments.

//----------------------------------------------------------------------------
// TAN (gnomonic) projection: celestial coordinates -> standard coordinates
//   crval: [ra0, dec0] in degrees
//   coord: [ra, dec] in degrees
//   Returns: [xi, eta] in degrees
//----------------------------------------------------------------------------
function tanProject(crval, coord) {
   var ra0  = crval[0] * Math.PI / 180.0;
   var dec0 = crval[1] * Math.PI / 180.0;
   var ra   = coord[0] * Math.PI / 180.0;
   var dec  = coord[1] * Math.PI / 180.0;

   var cosDec  = Math.cos(dec);
   var sinDec  = Math.sin(dec);
   var cosDec0 = Math.cos(dec0);
   var sinDec0 = Math.sin(dec0);
   var dRA     = ra - ra0;
   var cosDRA  = Math.cos(dRA);

   var D = sinDec0 * sinDec + cosDec0 * cosDec * cosDRA;
   if (D <= 0) {
      return null; // Cannot project (opposite hemisphere)
   }

   var xi  = (cosDec * Math.sin(dRA)) / D * (180.0 / Math.PI);
   var eta = (cosDec0 * sinDec - sinDec0 * cosDec * cosDRA) / D * (180.0 / Math.PI);

   return [xi, eta];
}

//----------------------------------------------------------------------------
// TAN inverse projection: standard coordinates -> celestial coordinates
//   crval: [ra0, dec0] in degrees
//   standard: [xi, eta] in degrees
//   Returns: [ra, dec] in degrees
//----------------------------------------------------------------------------
function tanDeproject(crval, standard) {
   var ra0  = crval[0] * Math.PI / 180.0;
   var dec0 = crval[1] * Math.PI / 180.0;
   var xi   = standard[0] * Math.PI / 180.0;
   var eta  = standard[1] * Math.PI / 180.0;

   var rho = Math.sqrt(xi * xi + eta * eta);

   if (rho === 0) {
      return [crval[0], crval[1]];
   }

   var c = Math.atan(rho);
   var cosC = Math.cos(c);
   var sinC = Math.sin(c);
   var cosDec0 = Math.cos(dec0);
   var sinDec0 = Math.sin(dec0);

   var dec = Math.asin(cosC * sinDec0 + eta * sinC * cosDec0 / rho);
   var ra  = ra0 + Math.atan2(xi * sinC, rho * cosDec0 * cosC - eta * sinDec0 * sinC);

   // Normalize RA to 0-360
   var raDeg = ra * 180.0 / Math.PI;
   while (raDeg < 0) raDeg += 360.0;
   while (raDeg >= 360.0) raDeg -= 360.0;

   return [raDeg, dec * 180.0 / Math.PI];
}

//----------------------------------------------------------------------------
// Angular separation (Vincenty formula)
//   coord1, coord2: [ra, dec] in degrees
//   Returns: angular separation in degrees
//----------------------------------------------------------------------------
function angularSeparation(coord1, coord2) {
   var ra1  = coord1[0] * Math.PI / 180.0;
   var dec1 = coord1[1] * Math.PI / 180.0;
   var ra2  = coord2[0] * Math.PI / 180.0;
   var dec2 = coord2[1] * Math.PI / 180.0;

   var dRA = ra2 - ra1;
   var cosDec1 = Math.cos(dec1);
   var sinDec1 = Math.sin(dec1);
   var cosDec2 = Math.cos(dec2);
   var sinDec2 = Math.sin(dec2);

   var num1 = cosDec2 * Math.sin(dRA);
   var num2 = cosDec1 * sinDec2 - sinDec1 * cosDec2 * Math.cos(dRA);
   var den  = sinDec1 * sinDec2 + cosDec1 * cosDec2 * Math.cos(dRA);

   return Math.atan2(Math.sqrt(num1 * num1 + num2 * num2), den) * 180.0 / Math.PI;
}

//----------------------------------------------------------------------------
// Linear algebra utilities
//----------------------------------------------------------------------------

// Solve a linear system Ax = b using Gaussian elimination with partial pivoting
// A: n×n 2D array, b: length-n array
// Returns: solution vector, or null if singular
function solveLinearSystem(A, b) {
   var n = b.length;
   // Create augmented matrix [A|b]
   var aug = [];
   for (var i = 0; i < n; i++) {
      aug[i] = [];
      for (var j = 0; j < n; j++) {
         aug[i][j] = A[i][j];
      }
      aug[i][n] = b[i];
   }

   // Forward elimination with partial pivoting
   for (var col = 0; col < n; col++) {
      var maxVal = Math.abs(aug[col][col]);
      var maxRow = col;
      for (var row = col + 1; row < n; row++) {
         if (Math.abs(aug[row][col]) > maxVal) {
            maxVal = Math.abs(aug[row][col]);
            maxRow = row;
         }
      }
      if (maxVal < 1e-15) return null;

      if (maxRow !== col) {
         var tmp = aug[col];
         aug[col] = aug[maxRow];
         aug[maxRow] = tmp;
      }

      for (var row = col + 1; row < n; row++) {
         var factor = aug[row][col] / aug[col][col];
         for (var j = col; j <= n; j++) {
            aug[row][j] -= factor * aug[col][j];
         }
      }
   }

   // Back substitution
   var x = [];
   for (var i = n - 1; i >= 0; i--) {
      if (Math.abs(aug[i][i]) < 1e-15) return null;
      var sum = aug[i][n];
      for (var j = i + 1; j < n; j++) {
         sum -= aug[i][j] * x[j];
      }
      x[i] = sum / aug[i][i];
   }

   return x;
}

// Solve underdetermined system D*x = b (m < n) via minimum-norm solution
// x = D^T * (D * D^T)^{-1} * b
// D: m×n (m < n), b: length-m → returns length-n solution, or null if singular
function solveMinNorm(D, b) {
   var m = D.length;
   var n = D[0].length;

   // Try without regularization, then with Tikhonov regularization as fallback
   var y = null;
   for (var attempt = 0; attempt < 2; attempt++) {
      var G = [];
      for (var i = 0; i < m; i++) {
         G[i] = [];
         for (var j = 0; j < m; j++) {
            var s = 0;
            for (var k = 0; k < n; k++) s += D[i][k] * D[j][k];
            G[i][j] = s;
         }
      }
      if (attempt === 1) {
         var maxDiag = 0;
         for (var i = 0; i < m; i++) {
            if (G[i][i] > maxDiag) maxDiag = G[i][i];
         }
         var eps = maxDiag * 1e-10;
         for (var i = 0; i < m; i++) G[i][i] += eps;
      }
      y = solveLinearSystem(G, b);
      if (y !== null) break;
   }
   if (y === null) return null;

   // x = D^T * y
   var x = [];
   for (var j = 0; j < n; j++) {
      var s = 0;
      for (var i = 0; i < m; i++) s += D[i][j] * y[i];
      x[j] = s;
   }
   return x;
}

//----------------------------------------------------------------------------
// WCSFitter: Fit a TAN projection WCS from star pairs
//
//   starPairs: array of [{px, py, ra, dec, name}] (requires 3 or more)
//   imageWidth, imageHeight: image dimensions in pixels
//----------------------------------------------------------------------------
function WCSFitter(starPairs, imageWidth, imageHeight) {
   this.stars = starPairs;
   this.width = imageWidth;
   this.height = imageHeight;
   // CRPIX is the image center (FITS 1-based)
   this.crpix1 = imageWidth / 2.0 + 0.5;
   this.crpix2 = imageHeight / 2.0 + 0.5;
}

WCSFitter.prototype.solve = function () {
   var stars = this.stars;
   var nStars = stars.length;

   if (nStars < 3) {
      return {
         success: false,
         message: "At least 3 star pairs required (current: " + nStars + ")"
      };
   }

   // RA/DEC range check
   for (var i = 0; i < nStars; i++) {
      if (stars[i].ra < 0 || stars[i].ra >= 360) {
         return {
            success: false,
            message: "Star " + (i + 1) + " RA is out of range: " + stars[i].ra
         };
      }
      if (stars[i].dec < -90 || stars[i].dec > 90) {
         return {
            success: false,
            message: "Star " + (i + 1) + " DEC is out of range: " + stars[i].dec
         };
      }
   }

   var crpix1 = this.crpix1;
   var crpix2 = this.crpix2;

   // --- 1. CRVAL initial value = centroid of star celestial coordinates ---
   // Use 3D unit vector mean on the celestial sphere.
   // This correctly handles circumpolar fields where stars wrap around in RA.
   var sumVX = 0, sumVY = 0, sumVZ = 0;
   for (var i = 0; i < nStars; i++) {
      var raRad = stars[i].ra * Math.PI / 180.0;
      var decRad = stars[i].dec * Math.PI / 180.0;
      sumVX += Math.cos(decRad) * Math.cos(raRad);
      sumVY += Math.cos(decRad) * Math.sin(raRad);
      sumVZ += Math.sin(decRad);
   }
   var crval1 = Math.atan2(sumVY, sumVX) * 180.0 / Math.PI;
   if (crval1 < 0) crval1 += 360.0;
   var rXY = Math.sqrt(sumVX * sumVX + sumVY * sumVY);
   var crval2 = Math.atan2(sumVZ, rXY) * 180.0 / Math.PI;

   // --- 2-4. Iterate: TAN projection -> CD matrix fit -> CRVAL update ---
   var cd = [[0, 0], [0, 0]];
   var maxIter = 15;

   for (var iter = 0; iter < maxIter; iter++) {
      var crval = [crval1, crval2];

      // Compute standard coordinates via TAN projection
      var projOk = true;
      var xiArr = [];
      var etaArr = [];
      for (var i = 0; i < nStars; i++) {
         var proj = tanProject(crval, [stars[i].ra, stars[i].dec]);
         if (proj === null) {
            projOk = false;
            break;
         }
         xiArr.push(proj[0]);
         etaArr.push(proj[1]);
      }

      if (!projOk) {
         return {
            success: false,
            message: "TAN projection failed (stars may be in the opposite hemisphere)"
         };
      }

      // Pixel offset u, v (relative to CRPIX)
      // Standard FITS coordinate system: y=1 is at the image bottom. fits_y = height - py.
      var uArr = [];
      var vArr = [];
      for (var i = 0; i < nStars; i++) {
         uArr.push((stars[i].px + 1.0) - crpix1);
         vArr.push((this.height - stars[i].py) - crpix2);
      }

      // Compute terms of the normal equations
      var sumUU = 0, sumUV = 0, sumVV = 0;
      var sumUXi = 0, sumVXi = 0;
      var sumUEta = 0, sumVEta = 0;
      for (var i = 0; i < nStars; i++) {
         sumUU += uArr[i] * uArr[i];
         sumUV += uArr[i] * vArr[i];
         sumVV += vArr[i] * vArr[i];
         sumUXi  += uArr[i] * xiArr[i];
         sumVXi  += vArr[i] * xiArr[i];
         sumUEta += uArr[i] * etaArr[i];
         sumVEta += vArr[i] * etaArr[i];
      }

      // Solve CD matrix using Cramer's rule
      var det = sumUU * sumVV - sumUV * sumUV;
      if (Math.abs(det) < 1e-30) {
         return {
            success: false,
            message: "Normal equation determinant is zero (stars may be collinear)"
         };
      }

      cd[0][0] = (sumUXi * sumVV - sumVXi * sumUV) / det;   // CD1_1
      cd[0][1] = (sumUU * sumVXi - sumUV * sumUXi) / det;   // CD1_2
      cd[1][0] = (sumUEta * sumVV - sumVEta * sumUV) / det;  // CD2_1
      cd[1][1] = (sumUU * sumVEta - sumUV * sumUEta) / det;  // CD2_2

      // Update CRVAL: inverse transform CRPIX (image center) -> celestial coords
      // Since the offset at CRPIX is (0, 0), standard coords are also (0, 0)
      // -> CRVAL doesn't change. But if CRPIX is offset from the pixel center, update it.
      // Here we use a fixed CRPIX with fine-tuned CRVAL:
      // Correct CRVAL using the centroid of all star residuals
      var sumDXi = 0, sumDEta = 0;
      for (var i = 0; i < nStars; i++) {
         var predXi  = cd[0][0] * uArr[i] + cd[0][1] * vArr[i];
         var predEta = cd[1][0] * uArr[i] + cd[1][1] * vArr[i];
         sumDXi  += xiArr[i] - predXi;
         sumDEta += etaArr[i] - predEta;
      }
      var meanDXi  = sumDXi / nStars;
      var meanDEta = sumDEta / nStars;

      // Inverse transform residual centroid to celestial coords and update CRVAL
      var newCrval = tanDeproject([crval1, crval2], [meanDXi, meanDEta]);

      // Verify that updated CRVAL doesn't break TAN projection for any star.
      // For wide-field images, the non-linear TAN projection can cause the
      // CRVAL update to overshoot, pushing edge stars beyond the 90-degree limit.
      var updateOk = true;
      for (var j = 0; j < nStars; j++) {
         if (tanProject(newCrval, [stars[j].ra, stars[j].dec]) === null) {
            updateOk = false;
            break;
         }
      }
      if (updateOk) {
         var crval2Rad = crval2 * Math.PI / 180.0;
         var crvalDelta = Math.sqrt(
            Math.pow((newCrval[0] - crval1) * Math.cos(crval2Rad), 2) +
            Math.pow(newCrval[1] - crval2, 2)
         ) * 3600.0;  // degrees -> arcsec
         crval1 = newCrval[0];
         crval2 = newCrval[1];
         if (crvalDelta < 1e-4) break;  // < 0.0001 arcsec で収束
      }
   }

   // --- 5. Compute TAN-only residuals ---
   var crval = [crval1, crval2];
   var residuals = [];
   var totalResidSq = 0;

   for (var i = 0; i < nStars; i++) {
      var u = (stars[i].px + 1.0) - crpix1;
      var v = (this.height - stars[i].py) - crpix2;

      var predXi  = cd[0][0] * u + cd[0][1] * v;
      var predEta = cd[1][0] * u + cd[1][1] * v;
      var predCoord = tanDeproject(crval, [predXi, predEta]);
      var resid = angularSeparation([stars[i].ra, stars[i].dec], predCoord);
      var residArcsec = resid * 3600.0;
      residuals.push({
         name: stars[i].name || ("Star " + (i + 1)),
         residual_arcsec: residArcsec
      });
      totalResidSq += residArcsec * residArcsec;
   }

   var rmsArcsec = Math.sqrt(totalResidSq / nStars);

   // Pixel scale computation (from CD matrix singular values)
   var pixelScaleArcsec = Math.sqrt(Math.abs(cd[0][0] * cd[1][1] - cd[0][1] * cd[1][0])) * 3600.0;

   return {
      success: true,
      crval1: crval1,
      crval2: crval2,
      crpix1: crpix1,
      crpix2: crpix2,
      cd: cd,
      pixelScale_arcsec: pixelScaleArcsec,
      rms_arcsec: rmsArcsec,
      residuals: residuals,
      message: "WCS fit succeeded (RMS: " + rmsArcsec.toFixed(2) + " arcsec, "
         + "pixel scale: " + pixelScaleArcsec.toFixed(3) + " arcsec/px)"
   };
};

//----------------------------------------------------------------------------
// Sky to pixel conversion: RA/DEC -> pixel coordinates using WCS result
//   ra, dec: celestial coordinates in degrees
//   wcsResult: object with crval1, crval2, crpix1, crpix2, cd (2x2 matrix)
//   imageHeight: image height in pixels
//   Returns: {px, py} in 0-based PixInsight coordinates, or null if on opposite hemisphere
//----------------------------------------------------------------------------
function skyToPixel(ra, dec, wcsResult, imageHeight) {
   var proj = tanProject([wcsResult.crval1, wcsResult.crval2], [ra, dec]);
   if (!proj) return null;

   var xi = proj[0];
   var eta = proj[1];

   // CD inverse matrix
   var cd = wcsResult.cd;
   var det = cd[0][0] * cd[1][1] - cd[0][1] * cd[1][0];
   if (Math.abs(det) < 1e-30) return null;

   var u = (cd[1][1] * xi - cd[0][1] * eta) / det;
   var v = (-cd[1][0] * xi + cd[0][0] * eta) / det;

   // FITS -> PixInsight coordinate conversion
   var px = u + wcsResult.crpix1 - 1;
   var py = imageHeight - (v + wcsResult.crpix2);

   return { px: px, py: py };
}

//----------------------------------------------------------------------------
// Centroid computation (intensity-weighted center of gravity)
//
// Uses Image.sample(x, y, channel) in the PJSR environment.
// This function is PJSR-only.
//
//   image: PixInsight Image object
//   cx, cy: click position (0-based pixel coordinates)
//   radius: search radius in pixels (default 10)
//   Returns: {x, y} sub-pixel star center, or null on failure
//----------------------------------------------------------------------------
function computeCentroid(image, cx, cy, radius) {
   if (typeof radius === "undefined") radius = 10;

   var x0 = Math.max(0, Math.round(cx) - radius);
   var y0 = Math.max(0, Math.round(cy) - radius);
   var x1 = Math.min(image.width - 1, Math.round(cx) + radius);
   var y1 = Math.min(image.height - 1, Math.round(cy) + radius);

   // Use channel 0 (monochrome or R channel)
   var ch = 0;

   // Collect pixel values within the window for background estimation (median)
   var values = [];
   for (var y = y0; y <= y1; y++) {
      for (var x = x0; x <= x1; x++) {
         values.push(image.sample(x, y, ch));
      }
   }

   if (values.length === 0) return null;

   // Use median as background level
   values.sort(function (a, b) { return a - b; });
   var median = values[Math.floor(values.length / 2)];

   // Intensity-weighted centroid (background subtracted)
   var sumW = 0, sumWX = 0, sumWY = 0;
   for (var y = y0; y <= y1; y++) {
      for (var x = x0; x <= x1; x++) {
         var val = image.sample(x, y, ch) - median;
         if (val > 0) {
            sumW  += val;
            sumWX += val * x;
            sumWY += val * y;
         }
      }
   }

   if (sumW <= 0) return null;

   return {
      x: sumWX / sumW,
      y: sumWY / sumW
   };
}

// Export for Node.js environment (ignored in PJSR)
if (typeof module !== "undefined") {
   module.exports = {
      tanProject: tanProject,
      tanDeproject: tanDeproject,
      angularSeparation: angularSeparation,
      solveLinearSystem: solveLinearSystem,
      solveMinNorm: solveMinNorm,
      WCSFitter: WCSFitter,
      computeCentroid: computeCentroid,
      skyToPixel: skyToPixel
   };
}

File: python/core/wcs_math.py
"""WCS 数学関数モジュール — TAN投影、角距離、WCSフィッティング。

javascript/wcs_math.js の Python 1:1 移植。
"""

from __future__ import annotations

import math
from typing import Optional


def tan_project(
    crval: tuple[float, float], coord: tuple[float, float]
) -> Optional[tuple[float, float]]:
    """TAN (gnomonic) 正投影: 天球座標 → 標準座標 (xi, eta)。

    Parameters
    ----------
    crval : (ra0, dec0) 投影中心 [度]
    coord : (ra, dec) 投影対象 [度]

    Returns
    -------
    (xi, eta) [度] または反対半球の場合 None
    """
    ra0 = math.radians(crval[0])
    dec0 = math.radians(crval[1])
    ra = math.radians(coord[0])
    dec = math.radians(coord[1])

    cos_dec = math.cos(dec)
    sin_dec = math.sin(dec)
    cos_dec0 = math.cos(dec0)
    sin_dec0 = math.sin(dec0)
    d_ra = ra - ra0
    cos_dra = math.cos(d_ra)

    D = sin_dec0 * sin_dec + cos_dec0 * cos_dec * cos_dra
    if D <= 0:
        return None  # 投影不可（反対半球）

    xi = (cos_dec * math.sin(d_ra)) / D * math.degrees(1.0)
    eta = (cos_dec0 * sin_dec - sin_dec0 * cos_dec * cos_dra) / D * math.degrees(1.0)

    return (xi, eta)


def tan_deproject(
    crval: tuple[float, float], standard: tuple[float, float]
) -> tuple[float, float]:
    """TAN (gnomonic) 逆投影: 標準座標 (xi, eta) → 天球座標。

    Parameters
    ----------
    crval : (ra0, dec0) 投影中心 [度]
    standard : (xi, eta) 標準座標 [度]

    Returns
    -------
    (ra, dec) [度]  RA は [0, 360) に正規化
    """
    ra0 = math.radians(crval[0])
    dec0 = math.radians(crval[1])
    xi = math.radians(standard[0])
    eta = math.radians(standard[1])

    rho = math.sqrt(xi * xi + eta * eta)

    if rho == 0:
        return (crval[0], crval[1])

    c = math.atan(rho)
    cos_c = math.cos(c)
    sin_c = math.sin(c)
    cos_dec0 = math.cos(dec0)
    sin_dec0 = math.sin(dec0)

    dec = math.asin(cos_c * sin_dec0 + eta * sin_c * cos_dec0 / rho)
    ra = ra0 + math.atan2(xi * sin_c, rho * cos_dec0 * cos_c - eta * sin_dec0 * sin_c)

    # RA を 0-360 に正規化
    ra_deg = math.degrees(ra)
    ra_deg %= 360.0
    if ra_deg < 0:
        ra_deg += 360.0

    return (ra_deg, math.degrees(dec))


def angular_separation(
    coord1: tuple[float, float], coord2: tuple[float, float]
) -> float:
    """2点間の角距離（Vincenty 公式）。

    Parameters
    ----------
    coord1 : (ra1, dec1) [度]
    coord2 : (ra2, dec2) [度]

    Returns
    -------
    角距離 [度]
    """
    ra1 = math.radians(coord1[0])
    dec1 = math.radians(coord1[1])
    ra2 = math.radians(coord2[0])
    dec2 = math.radians(coord2[1])

    d_ra = ra2 - ra1
    cos_dec1 = math.cos(dec1)
    sin_dec1 = math.sin(dec1)
    cos_dec2 = math.cos(dec2)
    sin_dec2 = math.sin(dec2)

    num1 = cos_dec2 * math.sin(d_ra)
    num2 = cos_dec1 * sin_dec2 - sin_dec1 * cos_dec2 * math.cos(d_ra)
    den = sin_dec1 * sin_dec2 + cos_dec1 * cos_dec2 * math.cos(d_ra)

    return math.degrees(math.atan2(math.sqrt(num1 * num1 + num2 * num2), den))


class WCSFitter:
    """星ペアからアフィン WCS (TAN + CD行列) をフィッティングする。

    Parameters
    ----------
    star_pairs : list[dict]
        各要素は {"px": float, "py": float, "ra": float, "dec": float, "name": str}
        px, py は 0-based ピクセル座標。
    image_width : int
    image_height : int
    """

    def __init__(
        self,
        star_pairs: list[dict],
        image_width: int,
        image_height: int,
    ) -> None:
        self.stars = star_pairs
        self.width = image_width
        self.height = image_height
        self.crpix1 = image_width / 2.0 + 0.5
        self.crpix2 = image_height / 2.0 + 0.5

    def solve(self) -> dict:
        """WCS フィットを実行して結果辞書を返す。"""
        stars = self.stars
        n_stars = len(stars)

        if n_stars < 4:
            return {
                "success": False,
                "message": f"最低4つの星ペアが必要です（現在: {n_stars}）",
            }

        # RA/DEC の範囲チェック
        for i, s in enumerate(stars):
            if s["ra"] < 0 or s["ra"] >= 360:
                return {
                    "success": False,
                    "message": f"星 {i + 1} の RA が範囲外です: {s['ra']}",
                }
            if s["dec"] < -90 or s["dec"] > 90:
                return {
                    "success": False,
                    "message": f"星 {i + 1} の DEC が範囲外です: {s['dec']}",
                }

        crpix1 = self.crpix1
        crpix2 = self.crpix2

        # --- 1. CRVAL 初期値 = 星の天球座標重心 ---
        # RA はラップアラウンドを考慮してベクトル平均
        sum_cos_ra = 0.0
        sum_sin_ra = 0.0
        crval2 = 0.0
        for s in stars:
            ra_rad = math.radians(s["ra"])
            sum_cos_ra += math.cos(ra_rad)
            sum_sin_ra += math.sin(ra_rad)
            crval2 += s["dec"]

        crval1 = math.degrees(math.atan2(sum_sin_ra, sum_cos_ra))
        if crval1 < 0:
            crval1 += 360.0
        crval2 /= n_stars

        # --- 2-4. 反復: TAN投影 → CD行列フィット → CRVAL更新 ---
        cd = [[0.0, 0.0], [0.0, 0.0]]
        max_iter = 5

        for _iter in range(max_iter):
            crval = (crval1, crval2)

            # TAN投影で標準座標計算
            xi_arr: list[float] = []
            eta_arr: list[float] = []
            proj_ok = True
            for s in stars:
                proj = tan_project(crval, (s["ra"], s["dec"]))
                if proj is None:
                    proj_ok = False
                    break
                xi_arr.append(proj[0])
                eta_arr.append(proj[1])

            if not proj_ok:
                return {
                    "success": False,
                    "message": "TAN投影に失敗しました（星が反対半球にある可能性）",
                }

            # ピクセルオフセット u, v（CRPIX 基準）
            # X: px は 0-based 左起点、FITS も左起点 → fits_x = px + 1
            # Y: py は 0-based 上起点、FITS は下起点 → fits_y = height - py
            u_arr = [(s["px"] + 1.0) - crpix1 for s in stars]
            v_arr = [(self.height - s["py"]) - crpix2 for s in stars]

            # 正規方程式の各項を計算
            sum_uu = 0.0
            sum_uv = 0.0
            sum_vv = 0.0
            sum_u_xi = 0.0
            sum_v_xi = 0.0
            sum_u_eta = 0.0
            sum_v_eta = 0.0

            for i in range(n_stars):
                sum_uu += u_arr[i] * u_arr[i]
                sum_uv += u_arr[i] * v_arr[i]
                sum_vv += v_arr[i] * v_arr[i]
                sum_u_xi += u_arr[i] * xi_arr[i]
                sum_v_xi += v_arr[i] * xi_arr[i]
                sum_u_eta += u_arr[i] * eta_arr[i]
                sum_v_eta += v_arr[i] * eta_arr[i]

            # クレーメルの公式で CD 行列を解く
            det = sum_uu * sum_vv - sum_uv * sum_uv
            if abs(det) < 1e-30:
                return {
                    "success": False,
                    "message": "正規方程式の行列式がゼロです（星が一直線上にある可能性）",
                }

            cd[0][0] = (sum_u_xi * sum_vv - sum_v_xi * sum_uv) / det  # CD1_1
            cd[0][1] = (sum_uu * sum_v_xi - sum_uv * sum_u_xi) / det  # CD1_2
            cd[1][0] = (sum_u_eta * sum_vv - sum_v_eta * sum_uv) / det  # CD2_1
            cd[1][1] = (sum_uu * sum_v_eta - sum_uv * sum_u_eta) / det  # CD2_2

            # CRVAL 更新
            sum_d_xi = 0.0
            sum_d_eta = 0.0
            for i in range(n_stars):
                pred_xi = cd[0][0] * u_arr[i] + cd[0][1] * v_arr[i]
                pred_eta = cd[1][0] * u_arr[i] + cd[1][1] * v_arr[i]
                sum_d_xi += xi_arr[i] - pred_xi
                sum_d_eta += eta_arr[i] - pred_eta

            mean_d_xi = sum_d_xi / n_stars
            mean_d_eta = sum_d_eta / n_stars

            new_crval = tan_deproject((crval1, crval2), (mean_d_xi, mean_d_eta))
            crval1 = new_crval[0]
            crval2 = new_crval[1]

        # --- 5. 残差計算 ---
        crval = (crval1, crval2)
        residuals = []
        total_resid_sq = 0.0

        for i, s in enumerate(stars):
            u = (s["px"] + 1.0) - crpix1
            v = (self.height - s["py"]) - crpix2

            pred_xi = cd[0][0] * u + cd[0][1] * v
            pred_eta = cd[1][0] * u + cd[1][1] * v

            pred_coord = tan_deproject(crval, (pred_xi, pred_eta))

            resid = angular_separation((s["ra"], s["dec"]), pred_coord)
            resid_arcsec = resid * 3600.0
            residuals.append(
                {
                    "name": s.get("name", f"Star {i + 1}"),
                    "residual_arcsec": resid_arcsec,
                }
            )
            total_resid_sq += resid_arcsec * resid_arcsec

        rms_arcsec = math.sqrt(total_resid_sq / n_stars)
        pixel_scale_arcsec = (
            math.sqrt(abs(cd[0][0] * cd[1][1] - cd[0][1] * cd[1][0])) * 3600.0
        )

        return {
            "success": True,
            "crval1": crval1,
            "crval2": crval2,
            "crpix1": crpix1,
            "crpix2": crpix2,
            "cd": cd,
            "pixel_scale_arcsec": pixel_scale_arcsec,
            "rms_arcsec": rms_arcsec,
            "residuals": residuals,
            "message": (
                f"WCS フィット成功 (RMS: {rms_arcsec:.2f} arcsec, "
                f"ピクセルスケール: {pixel_scale_arcsec:.3f} arcsec/px)"
            ),
        }
